include/swift/Threading/TSan.h

```diff
@@ -0,0 +1,95 @@
+//===--- TSan.h - TSan support functions ---------------------- -*- C++ -*-===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2023 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+//
+// Helper functions for code that needs to integrate with the thread
+// sanitizer.  In particular, TSan can't see inside the runtime libraries,
+// so we occasionally need to give it a hint that we're doing synchronization
+// in order to avoid false positives.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef SWIFT_THREADING_TSAN_H
+#define SWIFT_THREADING_TSAN_H
+
+namespace swift {
+
+#if defined(_WIN32) || defined(__wasi__) || !__has_include(<dlfcn.h>)
+namespace tsan {
+
+inline bool enabled() { return false; }
+template <typename T> T *acquire(T *ptr) { return ptr; }
+template <typename T> T *release(T *ptr) { return ptr; }
+template <typename T> T *consume(T *ptr) { return ptr; }
+
+} // namespace tsan
+#else
+
+namespace threading_impl {
+
+extern bool tsan_enabled;
+extern void (*tsan_acquire)(const void *ptr);
+extern void (*tsan_release)(const void *ptr);
+
+} // namespace threading_impl
+
+namespace tsan {
+
+/// Returns true if TSan is enabled
+inline bool enabled() {
+  return threading_impl::tsan_enabled;
+}
+
+/// Indicate to TSan that an acquiring load has occurred on the current
```

Comment:
Same as my other comment.  Not sure about the wording to explain the APIs here.

Example: T1 & T2 contending on the same lock `x` (or using some other API to "synchronize" / establish order between them):
```
T1                | T2

write to y
tsan_release(x)
lock given away
      --> sync point -->
                     lock taken
                     tsan_aquire(x)
                     read from y
```

TSan now knows: T1 release(x) synchronized/happens before T2 acquire(x), access to y is well-ordered, don't diagnose race for accesses to y

Reply:
Having read this again this morning, I think we're describing the exact same thing; the point is that all writes that happened-before the `tsan_release(x)` are visible to the thread that did `tsan_acquire(x)`, thus the read is safe. I think the only difference is that the comments I added are describing things from the perspective of the C memory model rather than from the perspective of the internal implementation of TSan, but the fundamental point is that you need a `tsan_acquire()` to correspond to an acquiring load that TSan didn't see, and a `tsan_release()` to correspond to a releasing store that TSan didn't see.

The difference between _acquire_ and _consume_ has to do with carried dependencies; OTOH consume is weaker than acquire, the point being that you could use it to protect a structure to which you had a pointer, without forcing memory ordering constraints on anything that didn't involve an access through that pointer. Modelling it as a `tsan_acquire()` is kind of wrong, since on a CPU that does support consume, non-dependent reads of data written before the corresponding `tsan_release()` would technically be a bug.